docs/services.md

```diff
@@ -266,6 +266,17 @@ In order to allow users to choose a port number for their `Services`, we must
 ensure that no two `Services` can collide.  We do that by allocating each
 `Service` its own IP address.
 
+To ensure each service receives a unique IP, an internal allocator atomically
+updates a global allocation map in etcd prior to each service. The map object
+must exist in the registry for services to get IPs, otherwise creations will
+fail with a message indicating an IP could not be allocated. A background
+controller is responsible for creating that map (to migrate from older versions
+of Kubernetes that used in memory locking) as well as checking for invalid
+assignments due to administrator intervention and cleaning up any any IPs
+that were allocated but which no service currently uses. To alter the portal
+net range, the administrator must delete the service map key and restart the
```

Comment:
> On May 7, 2015, at 2:59 PM, Tim Hockin <notifications@github.com> wrote:
> 
> In docs/services.md:
> 
> > @@ -266,6 +266,17 @@ In order to allow users to choose a port number for their `Services`, we must
> >  ensure that no two `Services` can collide.  We do that by allocating each
> >  `Service` its own IP address.
> >  
> > +To ensure each service receives a unique IP, an internal allocator atomically
> > +updates a global allocation map in etcd prior to each service. The map object
> > +must exist in the registry for services to get IPs, otherwise creations will
> > +fail with a message indicating an IP could not be allocated. A background
> > +controller is responsible for creating that map (to migrate from older versions
> > +of Kubernetes that used in memory locking) as well as checking for invalid
> > +assignments due to administrator intervention and cleaning up any any IPs
> > +that were allocated but which no service currently uses. To alter the portal
> > +net range, the administrator must delete the service map key and restart the
> How would an admin do that - I thought there was no REST endpoint exposed?
> 
etcdctl.  It should never fail, but it's fail safe (ish)
> —
> Reply to this email directly or view it on GitHub.
>